pkg/ipamd/datastore/data_store.go

```diff
@@ -443,207 +445,113 @@ func (ds *DataStore) writeBackingStoreUnsafe() error {
-	_, ok = curENI.IPv4Prefixes[prefix.IP.String()]
-	if ok {
-		return errors.New(IPAlreadyInStoreError)
-	}
-	_, numIPsPerPrefix, _ := GetPrefixDelegationDefaults()
-	ds.total = ds.total + numIPsPerPrefix
-	ds.allocatedPrefix++
-	// Prometheus gauge
 	totalIPs.Set(float64(ds.total))
-
-	curENI.IPv4Prefixes[prefix.IP.String()] = &ENIPrefix{Prefix: prefix, AllocatedIPs: NewPrefixStore(numIPsPerPrefix), UsedIPs: 0, FreeIps: numIPsPerPrefix}
-	ds.log.Infof("Added ENI(%s)'s IPv4 Prefix %s to datastore", eniID, prefix.IP.String())
-	return nil
-}
-
-// DelIPv4AddressFromStore delete an IP of ENI from datastore
-func (ds *DataStore) DelIPv4AddressFromStore(eniID string, ipv4 string, force bool) error {
-	ds.lock.Lock()
-	defer ds.lock.Unlock()
-
-	curENI, ok := ds.eniPool[eniID]
-	if !ok {
-		return errors.New(UnknownENIError)
-	}
-
-	ipAddr, ok := curENI.IPv4Addresses[ipv4]
-	if !ok {
-		return errors.New(UnknownIPError)
-	}
-
-	if ipAddr.Assigned() {
-		if !force {
-			return errors.New(IPInUseError)
-		}
-		ds.log.Warnf("Force deleting assigned ip %s on eni %s", ipv4, eniID)
-		forceRemovedIPs.Inc()
-		ds.unassignPodIPv4AddressUnsafe(ipAddr)
-		if err := ds.writeBackingStoreUnsafe(); err != nil {
-			ds.log.Warnf("Unable to update backing store: %v", err)
-			// Continuing because 'force'
-		}
+	curENI.AvailableIPv4Cidrs[strIPv4Cidr] = &CidrInfo{
+		Cidr:          ipv4Cidr,
+		IPv4Addresses: make(map[string]*AddressInfo),
+		IsPrefix:      isPrefix,
 	}
-
-	ds.total--
-	// Prometheus gauge
-	totalIPs.Set(float64(ds.total))
-
-	delete(curENI.IPv4Addresses, ipv4)
-
-	ds.log.Infof("Deleted ENI(%s)'s IP %s from datastore", eniID, ipv4)
+	ds.log.Infof("Added ENI(%s)'s IP/Prefix %s to datastore", eniID, strIPv4Cidr)
 	return nil
 }
 
-// DelIPv4AddressFromStore delete an IP of ENI from datastore
-func (ds *DataStore) DelIPv4PrefixFromStore(eniID string, ipv4Prefix string, force bool) error {
+func (ds *DataStore) DelIPv4CidrFromStore(eniID string, cidr net.IPNet, force bool) error {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 
 	curENI, ok := ds.eniPool[eniID]
 	if !ok {
+		ds.log.Debugf("Unknown ENI %s while deleting the CIDR", eniID)
 		return errors.New(UnknownENIError)
 	}
+	strIPv4Cidr := cidr.String()
+	ipv4 := cidr.IP.String()
 
-	//split IP here
-	ip, ipnet, err := net.ParseCIDR(ipv4Prefix)
-	if err != nil {
-		return err
-	}
-
-	ipv4 := ip.String()
-	prefixlen := ipnet.Mask
-	ds.log.Debugf("Adding IP %s and prefix len %s", ipv4, prefixlen)
-
-	ipPrefix, ok := curENI.IPv4Prefixes[ipv4]
+	_, ok = curENI.AvailableIPv4Cidrs[strIPv4Cidr]
 	if !ok {
+		ds.log.Debugf("Unknown %s CIDR", strIPv4Cidr)
 		return errors.New(UnknownIPError)
 	}
 
-	if ipPrefix.UsedIPs != 0 {
-		if !force {
-			return errors.New(IPInUseError)
-		}
-		ds.log.Warnf("Force deleting assigned Prefix %s on eni %s", ipv4, eniID)
-		forceRemovedIPs.Inc()
-		//Cleanup datastore /32 ips
-		prefixDB := ipPrefix.AllocatedIPs
+	deletableCidr := curENI.AvailableIPv4Cidrs[strIPv4Cidr]
```

Comment:
why not `deletableCidr, ok := curENI.AvailableIPv4Cidrs[strIPv4Cidr]`

Reply:
Yeah, will fix this.